readme.md
# Readme

Readme file for gitlab repository root.

Note that the links in this particular document are compatible with gitlab.

## Documentation

There is a user manual and a sphinx-compiled documentation in [handybeam\doc\build\index.html](handybeam\doc\build\index.html). Most methods are documented fairly well. If anything is unclear, contact me on [slack](https://ultrahaptics.slack.com/team/UB0RDJ24B) using [@jurek](https://ultrahaptics.slack.com/team/UB0RDJ24B)

## Do use HandyBeam for:

* quick visualisation of example acoustic fields
* throw-away checks of what-if scenarios
* development of task-specific research code
* demonstrating that we are nice people (e.g. that we know what we are talking about, and that we are open-sourcing things)
* education, exploration, having fun
* cool animations and pleasant colours for decorating Your room.

## Do not use HandyBeam for:

* secrets ( instead, clone the repo and do Your development there. Do not push to the public repo. )
* winning a bad argument (there is that saying - if you want to lie, use statistics. This applies here too.)

## Be aware that:

* HandyBeam model makes **a lot** of assumptions, which might be an oversimplification of the problem that You are facing.

## Things that have been repeatedly asked for, but are waiting in the queue as of now

These are in their order in the queue:

* time domain propagation simulation
* advanced excitation solvers (bessel beam, sonotweezer)
* link to realtime LeapMotion & Dragonfly (visualize field on hand )
* reflections, scattering

## Links

## Legal & disclaimer

### License

----


Copyright 2019 Ultrahaptics 

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at [this location](http://www.apache.org/licenses/LICENSE-2.0)

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

----

### Disclaimer:

1. If something uncool happens, it's Your fault.
2. If it breaks, you get to keep both pieces.

----

Complying with the license, be notified that **HandyBeam is a descendant of cueBeam** which was developed between 2008-2018 in [University of Strathclyde](https://www.strath.ac.uk/research/subjects/electronicelectricalengineering/instituteforsensorssignalscommunications/centreforultrasonicengineering/). The original source code can be had from
[here.](https://github.com/CentreForUltrasonicEngineering/cueBeam_EngD)

Although the core technology is based on the same principles, Note that virtually no source code is shared with cueBEAM. For example, the cueBEAM has been written for Matlab, MEX and CUDA. HandyBeam is a complete rewrite in Python and OpenCl.

Note that as of release R1.0, compatibility with cueBEAM has been completely dropped.

----

## Support

Primary developer is Dr Jerzy Dziewierz. Contact me on [slack](https://ultrahaptics.slack.com/team/UB0RDJ24B) using [@jurek](https://ultrahaptics.slack.com/team/UB0RDJ24B)

## Contributors

in no particular order:

* Tom Carter - project sponsor
* Jerzy Dziewierz - primary developer and maintainer
* Orestis Gregoriou - legacy project sponsor
* Salvador Catsis - code contributor
* Rob Malkin - legacy code contributor
* Hugh Hopper, Joe Spillman, Adam Price - calibration data, legacy code (Acouistic Renderer), helpful discussions
* Brian Kappus - helpful discussions, multipoint code contributions
* Ben Long - helpful discussions
* University of Strathclyde - legacy code contributor

## Contributing

Note that as of release R1.0, compatibility with cueBEAM has been completely dropped.

Unless there are some UH rules that I do not know about, I'd say:

* For version number, use semantic versioning 2.0.0 [https://semver.org/](https://semver.org/])
* For new features and bugfixes, use [GitHub Flow model](https://guides.github.com/introduction/flow/)
    * ~~For branching workflow, use this: [http://nvie.com/posts/a-successful-git-branching-model/](http://nvie.com/posts/a-successful-git-branching-model/]).~~
    * ~~For new features, start a new folder, then a new submodule, and then use this: [http://tom.preston-werner.com/2010/08/23/readme-driven-development.html](http://tom.preston-werner.com/2010/08/23/readme-driven-development.html).~~
* For documentation, use this: [http://www.writethedocs.org/guide/writing/beginners-guide-to-docs/](http://www.writethedocs.org/guide/writing/beginners-guide-to-docs/)
* For anything else, ask [Jerzy Dziewierz](https://ultrahaptics.slack.com/team/UB0RDJ24B)

----

## Zen of HandyBeam

* *"There is no point in getting the wrong answer really, really fast"* -- speed of code matters, but only after it has been shown to operate correctly.

* *"All that we have is a model"* -  using simplifying assumptions is OK, just remember to describe the model in the documentation.

* *"Computer simulations are here to make the overall cost of doing the job cheaper, and not more expensive"*  (Richard O'Leary) -- Do not do time-consuming computations just because You can. Have a reason for spending Your time.

* *"Things should be as simple as possible, but not simpler"* (Albert Einstein)

* *"There are no unnatural things. There are only things that we do not know about nature"*

* *"Everyone knows that something is impossible to do. Then comes that new guy that doesn't know that. And he does it."* -- (Zygmunt Wrona, c.a. 1990). -- It is OK to experiment and have random whacky ideas. Also, it's OK to test the common knowledge and challenge authorities.

* *"Nothing is perfect"* -- it is OK to be wrong, as long as you invite the chance to get corrected.

### Note - all the points have their original authors, whom I sometimes cannot properly attribute. No claim is made to authorship of these.

----


changelog.md
# Change log and devel blog

## 2019-06-06 History cleanup

A new superproject structure is being set up:

in the root folder:
 
* `handybeam` repository contains core share-able core only
* `handybeam_extension` contains optional extensions to the core share-able code, to work on specific problems
* `handybeam_username` contains non-share-able code bits, figures and garbage
* `tmp` contains temporary files

## 2019-05-24 Documentation rush

* Jurek picks up development of the HandyBeam core
* This version has multiple contributions from Salvador Catsis
* Updating the Readme to prepare for open-source release 


## 2019-01-21 Documentation rush

* Lots of updates to documentation
    * the docstrings now have 100% coverage
* Trims to source code and minor refactorings
* Tests have not ran for a while now, I expect many things to be broken
* Starting to split the public code from add-on-development-only code. New domain: UH



## 2019-01-07 Development restart

* Lots of updates to documentation
	* Moving obsolete parts of the documentation to the [obsolete] folder
* Small trims of the source code as I work through documenting it
* New concept: higher performance, and more flexible TxArray objects
	* there will be many TxArray types, one for each special case
	* TxArray instance holds all the data about it's transducers, in a uniform float-typed memory block,  no need for list of TxElement objects. This should make the operations on the elements significantly faster. 
	* The TxElement becomes a 'viewer' onto the flat memory block of the TxArray
	* I can begin introducing this by simply making another TxArray Class - with new features
	* The propagators need a better architecture . . . .
* New concept: reference propagators
	* should make it easier to write tests
	* should make it easier to demonstrate equivalence and correctness
	* should be good for education


## 2018-10-04

* sampling point list sampler and core - debugging
* Studying User Stories with Simon

## 2018-10-03

* bugfix in the propagator.py
* sampling point list sampler and core - first implementation

## 2018-10-02

Bump version number to 1.0.5

* Learning about software engineering with Simon
  * The production process
  * The tests & coverage
  * Capturing the requirements - user stories
  * the V-model
* Refactoring field* and parent* to sampler*
* adding named parameters instead of positional parameters everywhere possible
* watching https://www.youtube.com/watch?v=OSGv2VnC0go - a way towards more beautiful python
  * lot's of advice there, I will want to use it, here are the more relevant ones:
    04:47 -- Looping over a collection
    06:51 -- Looping over a collection of indicies
    07:36 -- Looping over two collections
    21:10 -- Looping over dictionary keys and values
    31:10 -- Clarify function calls with keyword arguments
    32:17 -- Clarify multiple return values with named tuples
    33:13 -- Unpacking sequences
    34:01 -- Updating multiple state variables
    36:15 -- Simultaneous state updates
    38:24 -- Concatenating strings
    39:57 -- Using decorators to factor-out administrative logic
    41:19 -- Factor-out temporary contexts for decimal
    42:25 -- How to use locks
    46:04 -- Concise expressive one-liners

## 2018-10-01

* Created abstract_sampler_ortho and separated from abstract_sampler
* began renaming field* to sampler*:
  * handybeam.xy_plane.XYPlane is now handybeam.sampler_plane_xy.PlaneXY
  * handybeam.yz_plane.YZPlane is now handybeam.sampler_plane_yz.PlaneYZ
  * handybeam.world.add_plane is now handybeam.world.add_sampler
  * et cetera

## 2018-09-30

Added a documentation for the kernel cxyz.cl

Attempt to convert the documentation from markdown to ReStructuredText

## 2018-09-24

Improved bugcatcher version reporting. It will now report latest tag as version, and git repo commit SHA as the metadata.

HN-97 : the default phase directivity slope updated to zero to reflect the latest findings about how the phase behaves (Slack discussion from 2018-09-21)

HN-100 : example how to create a volume of data for display in Voreen

## 2018-09-19 branch: doc_generated, author: Jurek

creating the "moved subaperture" and "rotated subapertures" demo in `demo_rectilinear_hexa_sunflower_arrays.ipynb`

Added a new handy method : `handybeam.visualize.visualize_xy_and_array()` - to serve the needs of the demo above

## 2018-09-18 HandyBeam R1.0.0 release event

The first release.

handybeam/misc.py
"""
file :code:`misc.py`, module :py:mod:`handybeam.misc`

Functions and helpers that would not fit anywhere else.

Some of them might be awaiting refactoring or deletion.

"""


def copy_docstring(from_function, separator="\n", prepend=True):
    """ Decorator. Copies the docstring of `from_function` into the decorated function.

    This is used to make the documentation of the overriding methods nicer: enables to join the documentation of the super class and the current class in a single place.

    if :code:`prepend` is true, the current(local) function's goes first. Otherwise, the :code:`from_function`'s docstring goes first.

    Modified and extended from `https://stackoverflow.com/questions/13741998 <https://stackoverflow.com/questions/13741998/is-there-a-way-to-let-classes-inherit-the-documentation-of-their-superclass-with>`_
    """
    def _decorator(func):
        """ creates a new, modified function, to be returned instead of the original.

        Combines the docstrings of two functions: the decorated, and the one from  'from_function'
        """
        
        source_doc = from_function.__doc__

        if func.__doc__ is None:  # if there is no docstring at the local function
            func.__doc__ = source_doc
        else:  # if there is, connect the two
            if prepend:
                func.__doc__ = separator.join([func.__doc__, source_doc])
            else:
                func.__doc__ = separator.join([source_doc, func.__doc__, ])

        return func

    return _decorator


class HandyDict(dict):
    """

    Provides matlab-like setting/storage of items in a dict

    q=HandyDict()
    q.new_key = 'Hello world!'
    print(q.new_key)



    happily copypasted from https://stackoverflow.com/questions/2352181/how-to-use-a-dot-to-access-members-of-dictionary
    """
    def __init__(self, args={'True': True, 'False':False}):
        """ starts the instance.

        :param args: initial value in the dictionary. If not set, it initializes with a default value of :code:`{'True': True, 'False':False}`
        """
        super(HandyDict, self).__init__(args)
        if isinstance(args, dict):
            for k, v in args.items():
                if not isinstance(v, dict):
                    self[k] = v
                else:
                    self.__setattr__(k, HandyDict(v))

    def __getattr__(self, attr):
        """ responds to :code:`answer=dict.attribute`"""
        return self.get(attr)

    def __setattr__(self, key, value):
        """ responds to :code:`dict.key=value` """
        self.__setitem__(key, value)

    def __setitem__(self, key, value):
        """ responds to :code:`dict.key=value` """
        super(HandyDict, self).__setitem__(key, value)
        self.__dict__.update({key: value})

    def __delattr__(self, item):
        """ ? I don't know what this does."""
        self.__delitem__(item)

    def __delitem__(self, key):
        """ ? I don't know what this does."""
        super(HandyDict, self).__delitem__(key)
        del self.__dict__[key]


# pylint: disable = C0301
primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71,
          73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151,
          157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233,
          239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317,
          331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419,
          421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503,
          509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607,
          613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701,
          709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811,
          821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911,
          919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997, 1009, 1013,
          1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091,
          1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181,
          1187, 1193, 1201, 1213, 1217, 1223, 1229, 1231, 1237, 1249, 1259, 1277,
          1279, 1283, 1289, 1291, 1297, 1301, 1303, 1307, 1319, 1321, 1327, 1361,
          1367, 1373, 1381, 1399, 1409, 1423, 1427, 1429, 1433, 1439, 1447, 1451,
          1453, 1459, 1471, 1481, 1483, 1487, 1489, 1493, 1499, 1511, 1523, 1531,
          1543, 1549, 1553, 1559, 1567, 1571, 1579, 1583, 1597, 1601, 1607, 1609,
          1613, 1619, 1621, 1627, 1637, 1657, 1663, 1667, 1669, 1693, 1697, 1699,
          1709, 1721, 1723, 1733, 1741, 1747, 1753, 1759, 1777, 1783, 1787, 1789,
          1801, 1811, 1823, 1831, 1847, 1861, 1867, 1871, 1873, 1877, 1879, 1889,
          1901, 1907, 1913, 1931, 1933, 1949, 1951, 1973, 1979, 1987, 1993, 1997,
          1999, 2003, 2011, 2017, 2027, 2029, 2039, 2053, 2063, 2069, 2081, 2083,
          2087, 2089, 2099, 2111, 2113, 2129, 2131, 2137, 2141, 2143, 2153, 2161,
          2179, 2203, 2207, 2213, 2221, 2237, 2239, 2243, 2251, 2267, 2269, 2273,
          2281, 2287, 2293, 2297, 2309, 2311, 2333, 2339, 2341, 2347, 2351, 2357,
          2371, 2377, 2381, 2383, 2389, 2393, 2399, 2411, 2417, 2423, 2437, 2441,
          2447, 2459, 2467, 2473, 2477, 2503, 2521, 2531, 2539, 2543, 2549, 2551,
          2557, 2579, 2591, 2593, 2609, 2617, 2621, 2633, 2647, 2657, 2659, 2663,
          2671, 2677, 2683, 2687, 2689, 2693, 2699, 2707, 2711, 2713, 2719, 2729,
          2731, 2741, 2749, 2753, 2767, 2777, 2789, 2791, 2797, 2801, 2803, 2819,
          2833, 2837, 2843, 2851, 2857, 2861, 2879, 2887, 2897, 2903, 2909, 2917,
          2927, 2939, 2953, 2957, 2963, 2969, 2971, 2999, 3001, 3011, 3019, 3023,
          3037, 3041, 3049, 3061, 3067, 3079, 3083, 3089, 3109, 3119, 3121, 3137,
          3163, 3167, 3169, 3181, 3187, 3191, 3203, 3209, 3217, 3221, 3229, 3251,
          3253, 3257, 3259, 3271, 3299, 3301, 3307, 3313, 3319, 3323, 3329, 3331,
          3343, 3347, 3359, 3361, 3371, 3373, 3389, 3391, 3407, 3413, 3433, 3449,
          3457, 3461, 3463, 3467, 3469, 3491, 3499, 3511, 3517, 3527, 3529, 3533,
          3539, 3541, 3547, 3557, 3559, 3571, 3581, 3583, 3593, 3607, 3613, 3617,
          3623, 3631, 3637, 3643, 3659, 3671, 3673, 3677, 3691, 3697, 3701, 3709,
          3719, 3727, 3733, 3739, 3761, 3767, 3769, 3779, 3793, 3797, 3803, 3821,
          3823, 3833, 3847, 3851, 3853, 3863, 3877, 3881, 3889, 3907, 3911, 3917,
          3919, 3923, 3929, 3931, 3943, 3947, 3967, 3989, 4001, 4003, 4007, 4013,
          4019, 4021, 4027, 4049, 4051, 4057, 4073, 4079, 4091, 4093, 4099, 4111,
          4127, 4129, 4133, 4139, 4153, 4157, 4159, 4177, 4201, 4211, 4217, 4219,
          4229, 4231, 4241, 4243, 4253, 4259, 4261, 4271, 4273, 4283, 4289, 4297,
          4327, 4337, 4339, 4349, 4357, 4363, 4373, 4391, 4397, 4409, 4421, 4423,
          4441, 4447, 4451, 4457, 4463, 4481, 4483, 4493, 4507, 4513, 4517, 4519,
          4523, 4547, 4549, 4561, 4567, 4583, 4591, 4597, 4603, 4621, 4637, 4639,
          4643, 4649, 4651, 4657, 4663, 4673, 4679, 4691, 4703, 4721, 4723, 4729,
          4733, 4751, 4759, 4783, 4787, 4789, 4793, 4799, 4801, 4813, 4817, 4831,
          4861, 4871, 4877, 4889, 4903, 4909, 4919, 4931, 4933, 4937, 4943, 4951,
          4957, 4967, 4969, 4973, 4987, 4993, 4999, 5003, 5009, 5011, 5021, 5023,
          5039, 5051, 5059, 5077, 5081, 5087, 5099, 5101, 5107, 5113, 5119, 5147,
          5153, 5167, 5171, 5179, 5189, 5197, 5209, 5227, 5231, 5233, 5237, 5261,
          5273, 5279, 5281, 5297, 5303, 5309, 5323, 5333, 5347, 5351, 5381, 5387,
          5393, 5399, 5407, 5413, 5417, 5419, 5431, 5437, 5441, 5443, 5449, 5471,
          5477, 5479, 5483, 5501, 5503, 5507, 5519, 5521, 5527, 5531, 5557, 5563, 5569,
          5573, 5581, 5591, 5623, 5639, 5641, 5647, 5651, 5653, 5657, 5659, 5669, 5683,
          5689, 5693, 5701, 5711, 5717, 5737, 5741, 5743, 5749, 5779, 5783, 5791, 5801,
          5807, 5813, 5821, 5827, 5839, 5843, 5849, 5851, 5857, 5861, 5867, 5869, 5879,
          5881, 5897, 5903, 5923, 5927, 5939, 5953, 5981, 5987, 6007, 6011, 6029, 6037,
          6043, 6047, 6053, 6067, 6073, 6079, 6089, 6091, 6101, 6113, 6121, 6131, 6133,
          6143, 6151, 6163, 6173, 6197, 6199, 6203, 6211, 6217, 6221, 6229, 6247, 6257,
          6263, 6269, 6271, 6277, 6287, 6299, 6301, 6311, 6317, 6323, 6329, 6337, 6343,
          6353, 6359, 6361, 6367, 6373, 6379, 6389, 6397, 6421, 6427, 6449, 6451, 6469,
          6473, 6481, 6491, 6521, 6529, 6547, 6551, 6553, 6563, 6569, 6571, 6577, 6581,
          6599, 6607, 6619, 6637, 6653, 6659, 6661, 6673, 6679, 6689, 6691, 6701, 6703,
          6709, 6719, 6733, 6737, 6761, 6763, 6779, 6781, 6791, 6793, 6803, 6823, 6827,
          6829, 6833, 6841, 6857, 6863, 6869, 6871, 6883, 6899, 6907, 6911, 6917, 6947,
          6949, 6959, 6961, 6967, 6971, 6977, 6983, 6991, 6997, 7001, 7013, 7019, 7027,
          7039, 7043, 7057, 7069, 7079, 7103, 7109, 7121, 7127, 7129, 7151, 7159, 7177,
          7187, 7193, 7207, 7211, 7213, 7219, 7229, 7237, 7243, 7247, 7253, 7283, 7297,
          7307, 7309, 7321, 7331, 7333, 7349, 7351, 7369, 7393, 7411, 7417, 7433, 7451,
          7457, 7459, 7477, 7481, 7487, 7489, 7499, 7507, 7517, 7523, 7529, 7537, 7541,
          7547, 7549, 7559, 7561, 7573, 7577, 7583, 7589, 7591, 7603, 7607, 7621, 7639,
          7643, 7649, 7669, 7673, 7681, 7687, 7691, 7699, 7703, 7717, 7723, 7727, 7741,
          7753, 7757, 7759, 7789, 7793, 7817, 7823, 7829, 7841, 7853, 7867, 7873, 7877,
          7879, 7883, 7901, 7907, 7919]
""" List of primes. Useful for many things.

Prime numbers are innate to the world.

Example usage: :code:`this_value=handybeam.misc.primes[8]` returns :code:`19`

"""
# pylint: enable = C0301
